### ISSUE TITLE
Luby restarts follow more or less the inner-outer scheme

### ISSUE THREAD

**Issue by chanseokoh** [msoos/cryptominisat#236]
This is more of a design decision than an issue. However, at least there exists inconsistency between the Luby restart series and the geometric restart series.

The Luby series gets reset in each `Searcher::solve()` iteration (`loop_num` is reset to 0 every time), whereas the geometric series continues without interruption between solve iterations. That is, the Luby series currently works more or less similar to the inner-outer scheme.

````C++
    max_conflicts_this_restart = conf.restart_first;
    for(loop_num = 0
        ; !must_interrupt_asap()
````
... (omitted) ...
````C++
        switch (params.rest_type) {
            case restart_type_geom:
                max_conflicts_this_restart *= conf.restart_inc;
                break;
            case restart_type_luby:
                max_conflicts_this_restart = luby(conf.restart_inc, loop_num) * (double)conf.restart_first;
                break;
            default:
                break;
        }
````

**Comment by msoos** [comment]
Hey,

Thanks for reading through so carefully! Luckily, on line ~2006:

max_conflicts_this_restart = conf.restart_first;

So it gets reset too as far as I understand. I hope the above clears
things up!

Cheers,

Mate

**Comment by chanseokoh** [comment]
You're right. I failed to see that the geometric series is also being reset. Thanks!

Best,
Chanseok

**Comment by msoos** [comment]
Actually, while we are at it.. your system seems not to use luby. It uses a switching geometric/glue-based restart strategy. The geom starts at 100 and increments at 1.1 (but timeout is multiplied by 2x, so effectively starts at 200), and the glue is the standard glue. Am I right? I thought you were using luby/glue, but from the DIFF on your website it seems that it's geom/glue. Can you please tell me about this?

**Comment by msoos** [comment]
(see http://cs.nyu.edu/~chanseok/cominisatps/04_hybrid_restarts.patch )

**Comment by chanseokoh** [comment]
You're right. I don't use Luby. I never did. This issue that I opened was just an independent one (which turned out to be not an issue).

Precisely speaking, my system is no-restart/glue-restart. In the no-restart phase, restarts simply never happen.

The point here is to ensure having an extended period of time of no restarts from time to time. Luby has done the job quite nicely, and Luby/glue might also work well if done correctly. However, I took a simple approach first with no-restart/glue. I tried to mix Luby/glue some way but couldn't find a way to produce a better result. It's not trivial to find a right balance between Luby and glue. For now, no-restart/glue works very well.

Details:
The overall no-restart/glue-restart cycles follow the geometric series. The ratio between no-restart and glue-restart is 1 to 2, i.e., we spent twice more time in glue-restart. The solver starts with the no-restart phase until 100 conflicts (strictly enforced). It then switches to the glue-restart phase and runs for (roughly) 2 x 100 conflicts. This one cycle is then increased by 10%. So, the next no-restart phase runs for another 110 conflicts, after which the glue-restart phase runs for about 220 conflicts. Then 121 for no-restart, 242 for glue-restart, and so on.

A few more things:
1. I removed Glucose restart blocking in the glue phase. I think having the no-restart phase is enough.
2. Glue history (global and recent averages) is not collected in the no-restart phase because of the wildly different nature of the no-restart phase (glue values increase dramatically).

FYI, the ROKK solver in SAT-COMP'14 uses Luby/glue. It did a good job in the competition. The middle tier in the 3-tier learnt clause database is also inspired by ROKK.
